### PR TITLE
BUG: Don't raise for NDFrame.mask with nullable boolean

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes https://github.com/pandas-dev/pandas/issues/35429
- [x] closes https://github.com/pandas-dev/pandas/issues/36395
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

~(Need to wait until the v1.1.3 whatsnew gets merged before adding a release note.)~

This is taking the approach of treating NA as False within NDFrame.mask and NDFrame.where when dealing with a nullable boolean mask.

### CODE REVIEW

File: pandas/tests/frame/indexing/test_where.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def test_where_set(self, where_frame, float_string_frame):
 
         def _check_set(df, cond, check_dtypes=True):
             dfi = df.copy()
-            econd = cond.reindex_like(df).fillna(True)
+            econd = cond.reindex_like(df).fillna(False)
```
Comment (dsaxton) [inline pandas/tests/frame/indexing/test_where.py:162]: This test was broken by no longer setting fill_value = bool(inplace) so had to update

**Comment by simonjayhawkins** [comment]
> (Need to wait until the v1.1.3 whatsnew gets merged before adding a release note.)

ok to target 1.1.2 if ready before tagged

**Comment by dsaxton** [comment]
> ok to target 1.1.2 if ready before tagged

I can add a v1.1.2 note just in case

**Comment by dsaxton** [comment]
@simonjayhawkins I think having a GitHub Action for merge conflicts like you mentioned would be super useful, do you know by chance how that could be set up?

File: pandas/core/generic.py [hunk 2]
```diff
@@ -9094,9 +9093,12 @@ def mask(
         cond = com.apply_if_callable(cond, self)
 
         # see gh-21891
-        if not hasattr(cond, "__invert__"):
+        if not hasattr(cond, "__array__"):
```
Comment (jreback) [inline pandas/core/generic.py:9096]: what needed this?

File: pandas/core/generic.py [hunk 3]
```diff
@@ -9094,9 +9093,12 @@ def mask(
...
             cond = np.array(cond)
 
+        cond[isna(cond)] = False
```
Comment (jreback) [inline pandas/core/generic.py:9099]: why this?

**Review by jreback — CHANGES_REQUESTED** [review]
this is an intricate change and not appropriate for 1.1.2

File: pandas/core/generic.py [hunk 4]
```diff
@@ -8834,7 +8833,7 @@ def _where(
             # GH#21947 we have an empty DataFrame/Series, could be object-dtype
             cond = cond.astype(bool)
 
-        cond = -cond if inplace else cond
+        cond = ~cond if inplace else cond
```
[the remark this answers is not in the document]
Comment (dsaxton) [reply pandas/core/generic.py:8836]: This was also breaking because of nullable booleans (the test `test_where_nullable_boolean_mask` from `pandas/tests/frame/indexing/test_where.py` in this PR fails).

Currently we have (which may be another bug):
```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame(pd.array([True, False]))

In [3]: -df
Out[3]:
    0
0  -1
1   0

In [4]: ~df
Out[4]:
       0
0  False
1   True
```

It might be because we're going through an object dtype array:

```python
In [10]: -np.array([True, False], dtype=object)

Out[10]: array([-1, 0], dtype=object)
```

File: pandas/core/generic.py (hunk 2 above)
Comment (dsaxton) [reply pandas/core/generic.py:9096]: This was to make sure we have an array so we can index as below (e.g., it's possible we get here with a scalar and the block below breaks, for example)

File: pandas/core/generic.py (hunk 3 above)
Comment (dsaxton) [reply pandas/core/generic.py:9099]: I want to ensure we fillna with False before taking the inverse and _then_ filling with False (which is the same as filling with True here, which we don't want). If values are missing here we want them to be treated as False so that we don't apply the mask.

File: pandas/core/frame.py [hunk 5]
```diff
@@ -3122,6 +3122,8 @@ def _setitem_frame(self, key, value):
 
         self._check_inplace_setting(value)
         self._check_setitem_copy()
+        if not self._indexed_same(key):
+            key = key.reindex_like(self).fillna(False)
```
[the remark this answers is not in the document]
Comment (dsaxton) [reply pandas/core/frame.py:3126]: I did this because inside _where we're filling with False and so we end up incorrectly setting values that aren't selected by key (previously the correct fill_value seemingly was only used "by accident" since we had inplace=True and also fill_value = bool(inplace)).

**Comment by simonjayhawkins** [comment]
i've added the 1.1.3 milestone back while the change note is still in 1.1.3.rst

File: pandas/core/frame.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/frame.py:3126]: instead of this ok to add an arg to _where to do the inverting. *all* logic associated with the reindexing / filling must live there.

File: pandas/core/generic.py [hunk 6]
```diff
@@ -8835,7 +8834,7 @@ def _where(
...
```
Comment (jreback) [inline pandas/core/generic.py:8837]: if you are goin to change this then let's clean starting on L834.

**Review by jreback — CHANGES_REQUESTED** [review]
this patch needs to go in 1.2 as its a core change.

File: pandas/tests/series/indexing/test_mask.py [hunk 7]
```diff
@@ -25,11 +25,11 @@ def test_mask():
     s2 = -(s.abs())
     rs = s2.where(~cond[:3])
     rs2 = s2.mask(cond[:3])
-    tm.assert_series_equal(rs, rs2)
+    # tm.assert_series_equal(rs, rs2)
 
     rs = s2.where(~cond[:3], -s2)
     rs2 = s2.mask(cond[:3], -s2)
-    tm.assert_series_equal(rs, rs2)
+    # tm.assert_series_equal(rs, rs2)
```
Comment (dsaxton) [inline pandas/tests/series/indexing/test_mask.py:32]: Temporarily commented this and the test above which are failing. I think the expected output of these are incorrect, since when we have missing values in the mask (e.g. through misalignment), we can't expect the mask and where methods to be perfect inverses of each other.

That is, if we take the complement of the mask then we aren't doing anything to the NA values, so if we pass this to either method and both treat this as False, then we'll see different behavior. To me this is expected:

```python
>>> import numpy as np
>>> from pandas import Series
>>> 
>>> s = Series(np.random.randn(5))
>>> cond = Series([True, False, False, True, False], index=s.index)
>>> s2 = -(s.abs())
>>> mask = cond[:3]
>>> 
>>> print(s2)
0   -1.150047
1   -1.152933
2   -0.442946
3   -0.705565
4   -0.484288
dtype: float64
>>> print(mask)
0     True
1    False
2    False
dtype: bool
>>> print(s2.where(~mask))
0         NaN
1   -1.152933
2   -0.442946
3         NaN
4         NaN
dtype: float64
>>> print(s2.mask(mask))
0         NaN
1   -1.152933
2   -0.442946
3   -0.705565
4   -0.484288
dtype: float64
```

What do you think @jreback ?

File: pandas/core/generic.py [hunk 8]
```diff
@@ -8982,13 +8983,16 @@ def where(
 
             - 'raise' : allow exceptions to be raised.
             - 'ignore' : suppress exceptions. On error return original object.
-
         try_cast : bool, default False
             Try to cast the result back to the input type (if possible).
+        invert : bool, default False
+            Whether or not to invert `cond`. This is done after alignment and
+            filling any missing values with `False`.
```
Comment (simonjayhawkins) [inline pandas/core/generic.py:8990]: .. versionadded:: 1.2

and maybe a bit more about it (and an example?) in whatsnew.

File: pandas/tests/series/indexing/test_mask.py (hunk 7 above)
Comment (dsaxton) [reply pandas/tests/series/indexing/test_mask.py]: If I switch the expected to use invert=True the tests pass so doing that instead

File: doc/source/whatsnew/v1.2.0.rst [hunk 9]
```diff
@@ -96,6 +96,20 @@ For example:
...
+:meth:`Series.where` and :meth:`DataFrame.where` now include an argument ``invert`` to handle inversion of a possibly misaligned mask. The inversion is done after realignment so as to properly handle any missing values.
+
+.. ipython:: python
+
+    s = pd.Series(range(5), dtype=float)
+    mask = (s > 2)[:-1]
+    mask
+    s.where(mask)
+    s.where(~mask)
+    s.where(mask, invert=True)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:111]: oh woa, I wouldn't add to the public api at all, maybe my comment wasn't clear this should be for `_where` *only*

File: pandas/core/generic.py (hunk 8 above)
Comment (jreback) [reply pandas/core/generic.py:8993]: see my comments above, no need to add to the public api

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.2.0.rst (hunk 9 above)
Comment (dsaxton) [reply doc/source/whatsnew/v1.2.0.rst]: Moved this argument out of the public `where`.

I'm having to do a bit of boolean voodoo to get this to work. Had to bypass a call to `where` and go straight to `_where` which meant moving a small piece of logic that applies a callable "other" there as well. This broke some tests because apparently we sometimes we want to do this when using mask / where / setitem and other times not, so needed more control over whether this will happen.

**Comment by jbrockmendel** [comment]
seems like this is a workaround because BooleanArray doesn't have a `__inv__`.  would defining that solve this issue?

**Comment by jreback** [comment]
this is a good point @jbrockmendel

**Comment by dsaxton** [comment]
> seems like this is a workaround because BooleanArray doesn't have a `__inv__`. would defining that solve this issue?

Do you mean the issue with - vs ~? I might be misunderstanding but inverting a BooleanArray seems to work okay:

```python
[ins] In [3]: arr
Out[3]: 
<BooleanArray>
[True, False, <NA>]
Length: 3, dtype: boolean

[ins] In [4]: ~arr
Out[4]: 
<BooleanArray>
[False, True, <NA>]
Length: 3, dtype: boolean
```

This PR mostly became complicated because of the existing logic surrounding inplace inside _where, and when the inversion should happen in the presence of missing values.

File: pandas/core/frame.py [hunk 10]
```diff
@@ -3104,7 +3104,7 @@ def _setitem_frame(self, key, value):
 
         self._check_inplace_setting(value)
         self._check_setitem_copy()
-        self._where(-key, value, inplace=True)
+        self._where(key, value, inplace=True, invert=True)
```
Comment (jreback) [inline pandas/core/frame.py:3107]: so I would remove the invert keyword and just fix panda/core/series as well (and have _where *always* invert)

File: pandas/core/generic.py [hunk 11]
```diff
@@ -8857,13 +8857,18 @@ def _where(
         level=None,
         errors="raise",
         try_cast=False,
+        invert=False,
```
Comment (jreback) [inline pandas/core/generic.py:8860]: see my comments above

File: pandas/core/generic.py [hunk 12]
```diff
@@ -8857,13 +8857,18 @@ def _where(
...
+        apply_callable_other=False,
     ):
         """
         Equivalent to public method `where`, except that `other` is not
         applied as a function even if callable. Used in __setitem__.
         """
         inplace = validate_bool_kwarg(inplace, "inplace")
 
+        if apply_callable_other:
```
Comment (jreback) [inline pandas/core/generic.py:8869]: why is this needed? I am *almost* always -1 on adding args to anything to fix special cases.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by simonjayhawkins** [comment]
> this is an intricate change and not appropriate for 1.1.2

As this PR evolves, can we review the milestone when this PR is ready but before this is merged.

File: pandas/core/generic.py (hunk 12 above)
Comment (dsaxton) [reply pandas/core/generic.py:8869]: This was needed because evidently when we setitem using a mask and a callable other, we don't actually want that callable to get applied (and it passes through the `_where` code path):
```python
[ins] In [1]: import pandas as pd
         ...:
         ...: inc = lambda x: x + 1
         ...:
         ...: s = pd.Series([1, 2, -1, 4])
         ...: s[s < 0] = inc
         ...: s
         ...:
Out[1]:
0                                        1
1                                        2
2    <function <lambda> at 0x7fa7fb59ef70>
3                                        4
dtype: object
[ins] In [2]: pd.__version__
Out[2]: '1.1.2'
```
But other methods like `mask` and `where` need for any functions to be evaluated to booleans. Previously that logic existed outside `_where`, but if it's moved inside then I think some control is needed over whether or not to apply the function.

File: pandas/core/frame.py (hunk 10 above)
Comment (dsaxton) [reply pandas/core/frame.py:3107]: Hmm, I think it could be confusing if the default behavior of `_where` is to do essentially the opposite of what it's done previously.

This keyword is only to handle the special case of inverting after filling when dealing with missing values / misalignment, and while it's not pretty it's at least not user-facing. I'm not really aware of an elegant way to do this given that realignment is already happening inside `_where`.

File: doc/source/whatsnew/v1.2.0.rst [hunk 13]
```diff
@@ -305,6 +305,7 @@ Indexing
 Missing
 ^^^^^^^
 
+- Bug in :meth:`DataFrame.mask` and :meth:`DataFrame.where` raising an ``AssertionError`` when using a nullable boolean mask (:issue:`35429`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.2.0.rst:308]: > This is taking the approach of treating NA as False within NDFrame.mask and NDFrame.where when dealing with a nullable boolean mask.

hmm, thinking some more...

from https://pandas.pydata.org/pandas-docs/dev/reference/api/pandas.DataFrame.mask.html

> Where cond is False, keep the original value. Where True, replace with corresponding value from other.

and 

```
>>> bool(pd.NA)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas\_libs\missing.pyx", line 360, in pandas._libs.missing.NAType.__bool__
    raise TypeError("boolean value of NA is ambiguous")
TypeError: boolean value of NA is ambiguous
>>>
```

maybe this isn't a bug and this should raise, but not AssertionError, see https://github.com/pandas-dev/pandas/issues/35429#issuecomment-668087893

otherwise I think the docs should to be updated to be explicit about the behaviour.

We could be in danger here of introducing a behaviour that we would want to deprecate in the future.

File: pandas/tests/series/indexing/test_where.py [hunk 14]
```diff
@@ -452,3 +452,19 @@ def test_where_empty_series_and_empty_cond_having_non_bool_dtypes():
...
+    # https://github.com/pandas-dev/pandas/issues/35429
+    ser = Series([1, 2, 3])
+    mask = Series([True, False, None], dtype="boolean")
+    expected = Series([1, 999, 999])
+
+    if inplace:
+        result = ser.copy()
+        result.where(mask, 999, inplace=True)
+    else:
+        result = ser.where(mask, 999, inplace=False)
```
Comment (simonjayhawkins) [inline pandas/tests/series/indexing/test_where.py:468]: following on from previous comment.

```
>>> ser = pd.Series([1, 2, 3], dtype="Int64")
>>> mask = pd.Series([True, False, None], dtype="boolean")
>>> ser.where(mask, 999)
0      1
1    999
2    999
dtype: Int64
```
using a nullable mask with a nullable type gives the wrong output?

File: doc/source/whatsnew/v1.2.0.rst (hunk 13 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v1.2.0.rst:308]: i tend to agree with this

Comment (dsaxton) [reply doc/source/whatsnew/v1.2.0.rst:308]: Because mask / where are used in other places that would mean those other cases are left broken as well, e.g., https://github.com/pandas-dev/pandas/issues/36395. It also feels natural that `s[cond] += 999` be equivalent to `s = s.mask(cond, 999)` and on master:
```python
[ins] In [16]: s
Out[16]: 
0    1
1    2
2    3
dtype: int64

[ins] In [17]: cond
Out[17]: 
0     True
1    False
2     <NA>
dtype: boolean

[ins] In [18]: s[cond] = 999

[ins] In [19]: s
Out[19]: 
0    999
1      2
2      3
dtype: int64
```

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.2.0.rst:393]: The example in #36395 looks like its trying to use `np.array([[True, pd.NA]], dtype=object)` as a boolean mask, which i think _should_ raise (thought not AssertionError)

(with 2D EAs we can make this a little more nicely behaved and have `BooleanArray([[True, pd.NA]])`) and get a nicer exception message, but it still isn't clear that should be allowed

File: pandas/tests/series/indexing/test_where.py (hunk 14 above)
Comment (dsaxton) [reply pandas/tests/series/indexing/test_where.py:468]: This output is correct IMO, but the output changes if the operation is inplace:
```python
[ins] In [4]: s.where(mask, 999)
Out[4]: 
0      1
1    999
2    999
dtype: Int64

[ins] In [5]: s.where(mask, 999, inplace=True)

[ins] In [6]: s
Out[6]: 
0      1
1    999
2      3
dtype: Int64
```

File: doc/source/whatsnew/v1.2.0.rst (hunk 13 above)
Comment (dsaxton) [reply doc/source/whatsnew/v1.2.0.rst:393]: I think allowing indexing with nullable booleans is more useful than raising personally. This already works for single 1D arrays (the same way it would when filtering with a nullable boolean column in SQL):
```python
[ins] In [11]: s
Out[11]: 
0    1
1    2
2    3
dtype: Int64

[ins] In [12]: mask
Out[12]: 
0     True
1    False
2     <NA>
dtype: boolean

[ins] In [13]: s[mask]
Out[13]: 
0    1
dtype: Int64
```
The argument for allowing it for mask / where as well for me seems to be that these are very much like indexing operations, for which it's useful to treat as above.

File: pandas/core/frame.py (hunk 10 above)
Comment (jreback) [reply pandas/core/frame.py:3114]: _where is private so this doesn't matter, we want as consistent AND minimal api internally as possible.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 10 above)
Comment (dsaxton) [reply pandas/core/frame.py:3114]: I can try to get this to work, but changing _where globally in this way breaks a large number of tests

**Comment by dsaxton** [comment]
Closing for now since it seems it isn't clear if / how we want to do this
